---
title: "Hyperosmotic Gene Expression"
output:
  html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Hyperosmotic Gene Expression}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: osmotic_gene.bib
csl: peerj.csl
---

```{r setup, include=FALSE}
library(canprot)
library(CHNOSZ)
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available (R-Forge?)
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

This vignette from the R package [JMDplots](https://github.com/jedick/JMDplots) version `r packageDescription("JMDplots")$Version` shows compositional metrics and phylostrata for proteins coded by genes that are differentially expressed in hyperosmotic compared to control conditions.
Abbreviations:

  * <i>Z</i><sub>C</sub> &ndash; carbon oxidation state; <i>n</i><sub>H<sub>2</sub>O</sub> &ndash; stoichiometric hydration state per residue.
  * <i>n</i><sub>AA</sub> &ndash; protein length; MW &ndash; molecular weight per residue.

Stoichiometric values are calculated using basis species (rQEC derivation; [<span title="Distinct trends in chemical composition of proteins from metagenomes in redox and salinity gradients">Dick et al., 2020</span>](https://doi.org/10.1101/2020.04.01.020008); [<span title="Water as a reactant in the differential expression of proteins in cancer">Dick, 2020</a>](https://doi.org/10.1101/2020.04.09.035022)).

```{r options, echo=FALSE}
options(width = 90)
```

<style type="text/css">
body {
  max-width: 800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r datasets}
datasets <- pdat_osmotic_gene(2020)
```

```{r comptab, results="hide", message=FALSE, echo = FALSE}
pdat1 <- lapply(datasets, pdat_osmotic_gene)
comptab1 <- lapply(pdat1, get_comptab)
comptab2 <- lapply(pdat1, get_comptab, "pI", "GRAVY")
comptab3 <- lapply(pdat1, get_comptab, "nAA", "MW")
```

Differences are calculated as (median value for up-regulated proteins) - (median value for down-regulated proteins).
Dashed lines enclose the 50% credible region for highest probability density.

```{r diffplot, fig.width=9, fig.height=8, fig.align = "center", echo = FALSE, pngquant = pngquant}
layout(matrix(c(0, 2, 1, 2, 1, 3, 0, 3), nrow = 2))
par(cex = 1, mar = c(4, 4, 1, 1), mgp = c(2.5, 1, 0))
diffplot(comptab1, labtext = NA)
diffplot(comptab2, c("pI", "GRAVY"), labtext = NA)
diffplot(comptab3, c("nAA", "MW"), labtext = NA)
```

In the table, values of &Delta;<i>Z</i><sub>C</sub>, &Delta;<i>n</i><sub>H<sub>2</sub>O</sub>, and &Delta;GRAVY are multiplied by 1000, values of &Delta;pI and &Delta;MW are multiplied by 100, and negative values are shown in bold.
Abbreviations:

  * <i>n</i><sub>down</sub> &ndash; number of down-regulated proteins; <i>n</i><sub>up</sub> &ndash; number of up-regulated proteins.

```{r xsummary, results="asis", echo = FALSE}
library(xtable)
out <- xsummary3(comptab1, comptab2, comptab3)
# round values and include dataset tags
tags <- sapply(sapply(strsplit(datasets, "="), "[", -1), paste, collapse = ";")
out <- cbind(out[, 1:2], tags = tags, out[, 3:19])
out[, 6:20] <- round(out[, 6:20], 4)
write.csv(out, "osmotic_gene.csv", row.names = FALSE, quote = 2)
```

## Data Sources

__a__. __b__. Tables 1--2 of @KSA+02.
__c__. Table 1 of @WJ02.
__d__. __e__. Table 3 of @HZP+05.
__f__. Table 1S of @LGW+05.
__g__. Tables 2--3 of @SLA+05.
__h__. __i__. [Dataset.txt](http://www.wright.edu/~oleg.paliy/Papers/Osmo_Temp/Dataset.txt) of @GCP08.
__j__. __k__. Table 3 of @SBB+09.
__l__. Table S2 (sheet "S2_Table_3") of @HMO+10.
__m__. __n__. __o__. __p__. Table S5 of @BBWB12, filtered to include genes with log~2~ fold change > 1 or < -1.
__q__. Table S2 of @LB12, filtered to include genes with _p_-value < 0.05 and fold change > 2 or < 0.5.
__r__. __s__. __t__. Supplementary Tables S2A--S2F of @QHT+13.
__u__. __v__. [Dataset.txt](http://www.wright.edu/~oleg.paliy/Papers/UTI_Osmo/Dataset.txt) of @WGB+13, filtered to include proteins with log fold change (column "N vs K Wmean [M]" for NaCl and "U vs K Wmean [M]" for urea) > 1 or < 1 and _p_-value less than 0.05.
__w__. Table S2 of @ADW+14
__x__. __y__. __z__. Table S2 of @KKG+14
__A__. __B__. Tables S5--S6 of @KSM+14.
__C__. __D__. __E__. __F__. Supplemental Table S1 of @MGM+14, filtered to include genes with differences of Normalized ln(cDNA/gDNA) > 1 or < 1 between 2 % and higher NaCl concentration.
__G__. __H__. __I__. Table S2 of @SLM+14.
__J__. __K__. __L__. __M__. __N__. __O__. __P__. __Q__. __R__. Tables S1--S3 of @FRH+15.
__S__. __T__. Table S2 of @KLB+15.
__U__. __V__. Tables S5--S8 of @HLL17.
__W__. Table S6 of @MWZ+18.

## References
