---
title: "Prostate Cancer"
output:
  html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Prostate Cancer}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: cpdat.bib
csl: peerj.csl
---

```{r setup, include=FALSE}
library(JMDplots)
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available (R-Forge?)
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
# read 'basis' option
basis <- getOption("basis")
```

This vignette from the R package [JMDplots](https://github.com/jedick/JMDplots) version `r packageDescription("JMDplots")$Version` shows chemical metrics for proteins that are differentially expressed in prostate cancer compared to normal tissue.
The analysis is described in more detail in a paper ([<span title="Water as a reactant in the differential expression of proteins in cancer">Dick, 2021</a>](https://doi.org/10.1002/cso2.1007)).
Abbreviations:

  * <i>Z</i><sub>C</sub> &ndash; average oxidation state of carbon.
  * <i>n</i><sub>H<sub>2</sub>O</sub> &ndash; stoichiometric hydration state per residue ([<span title="Uncovering chemical signatures of salinity gradients through compositional analysis of protein sequences">Dick et al., 2020</span>](https://doi.org/10.5194/bg-17-6145-2020)) computed using the **`r basis`** basis species (`r basis.text(basis)`, H~2~O, O~2~).

<style type="text/css">
body {
  max-width: 800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r datasets}
datasets <- pdat_prostate(2020)
```

```{r comptab, results="hide", message=FALSE, echo = FALSE}
pdat1 <- lapply(datasets, pdat_prostate)
comptab1 <- lapply(pdat1, get_comptab)
comptab2 <- lapply(pdat1, get_comptab, "nAA", "MW")
```

Differences are calculated as (median value for up-regulated proteins) - (median value for down-regulated proteins).
Dashed lines enclose the 50% confidence region for highest probability density.

```{r diffplot, fig.width=5, fig.height=5, fig.align = "center", echo = FALSE, pngquant = pngquant}
par(mar = c(4, 4, 1, 1))
diffplot(comptab1, labtext = NA)
```

In the table, values of &Delta;<i>Z</i><sub>C</sub> and &Delta;<i>n</i><sub>H<sub>2</sub>O</sub> are multiplied by 1000, values of &Delta;MW are multiplied by 100, and negative values are shown in bold.
Abbreviations:

  * _n_~down~ &ndash; number of down-regulated proteins; _n_~up~ &ndash; number of up-regulated proteins; PCa &ndash; prostate cancer; BPH &ndash; benign prostatic hyperplasia; FFPE &ndash; formalin-fixed paraffin-embedded; OCT &ndash; optimal cutting temperature compound-embedded; TMA &ndash; tissue microarray; GS &ndash; Gleason Score; CiRT &ndash; common internal retention time.
  * <i>n</i><sub>AA</sub> &ndash; protein length; MW &ndash; molecular weight per residue.

```{r xsummary, results="asis", echo = FALSE}
library(xtable)
out <- xsummary2(comptab1, comptab2)
# round values and include dataset tags
tags <- sapply(sapply(strsplit(datasets, "="), "[", -1), paste, collapse = ";")
out <- cbind(out[, 1:2], tags = tags, out[, 3:16])
out[, 6:17] <- round(out[, 6:17], 4)
file <- paste0("prostate_", basis, ".csv")
write.csv(out, file, row.names = FALSE, quote = 2)
```

## Data Sources
Gene names or other identifiers were converted to UniProt accession numbers using the <a href="https://www.uniprot.org/mapping/">UniProt mapping tool</a>.

__a__. Table 2 of @GTR+08.
__b__. Supplementary Tables 3 and 4 of @KPB+10.
__c__. Table S2 of @HZH+12.
__d__. Table S1 of @JHZ+13.
__e__. Supplemental Table S7 of @LCS+14.
__f__. Supplementary Table 1 @CZL+16 (proteins recorded with only Down-regulation or Up-regulation).
__g__. Supplementary Table S3 of @IWT+16, filtered to include proteins listed with FDR < 0.1.
__h__. __i__. Extracted from Table S4 @GLZ+18. Values were quantile normalized, then ratios were calculated between the median values for each cancer type (acinar and ductal) and corresponding normal tissue; ratios > 1.5 or < 2/3 were used to identify differentially expressed proteins.
__j__. __k__. Extracted from Supplementary Data 1 of @LAJ+18 (sheet "Area-proteins") by applying quantile normalization to peak areas then calculating median values across all runs and samples for each of BPH , PC (primary prostate cancer), and CRPC (castration resistant prostate cancer). A cutoff of 2-fold in ratios of medians (PC / BPH or CRPC / BPH) was used to identify differentially expressed proteins.
__l__. Table 2 of @MAN+18.
__m__. __n__. __o__. __p__. __q__. Supporting Information Table S3 of @KRN+19 (G1--G5: PCa grades).
__r__. __s__. Tables S4d (GS = 6) and S4f (GS â‰¥ 8 or GS = 6) of @MMF+20 (GS: Gleason Score).
__t__. Supplementary Table S3a of @TOT+19.
__u__. __v__. Table S5 of @ZYW+19 (LG: low-grade PCa; HG: high-grade PCa).
__w__. Supplemental Table S2 of @KHN+20, filtered to include proteins with log~2~ fold change > 1 or < -1 in at least one experiment.
__x__. Table S2 of @LDM+20.
__y__. Table S2, Sheet D:S3-SWATH_protein_matrix of @SHC+20, filtered to include proteins quantified in at least 50% of both tumor and normal samples and with median fold change > 2 or < 0.5.
__z__. Table S4 of @ZKL+20 (*Pten* gene knockout vs wild-type).
__A__. Supplementary Tables S3E--S3F of @ZZX+20.

## References
