---
title: "Yeast Gene Expression in Hyper- and Hypo-osmotic Shock"
output:
  html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Yeast Gene Expression in Hyper- and Hypo-osmotic Shock}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: osmotic_gene.bib
csl: peerj.csl
---

```{r setup, include=FALSE}
library(canprot)
library(CHNOSZ)
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available (R-Forge?)
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
# read 'basis' option
basis <- getOption("basis")
```

This vignette from the R package [JMDplots](https://github.com/jedick/JMDplots) version `r packageDescription("JMDplots")$Version` shows compositional metrics for proteins coded by genes that are differentially expressed in hyperosmotic and hypoosmotic shock in yeast.
The analysis is described in more detail in a paper ([<span title="Water as a reactant in the differential expression of proteins in cancer">Dick, 2020</a>](https://doi.org/10.1002/cso2.1007)).
Abbreviations:

  * <i>Z</i><sub>C</sub> &ndash; average oxidation state of carbon.
  * <i>n</i><sub>H<sub>2</sub>O</sub> &ndash; stoichiometric hydration state per residue ([<span title="Uncovering chemical signatures of salinity gradients through compositional analysis of protein sequences">Dick et al., 2020</span>](https://doi.org/10.5194/bg-17-6145-2020)) computed using the **`r basis`** basis species (`r basis.text(basis)`, H~2~O, O~2~).
  * <i>n</i><sub>AA</sub> &ndash; protein length; MW &ndash; molecular weight per residue.

```{r options, echo=FALSE}
options(width = 90)
```

<style type="text/css">
body {
  max-width: 800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r datasets}
datasets <- pdat_yeast_stress(2020)
```

```{r comptab, results="hide", message=FALSE, echo = FALSE}
pdat1 <- lapply(datasets, pdat_yeast_stress)
comptab1 <- lapply(pdat1, get_comptab)
comptab2 <- lapply(pdat1, get_comptab, "pI", "GRAVY")
comptab3 <- lapply(pdat1, get_comptab, "nAA", "MW")
```

Differences are calculated as (median value for proteins coded by up-regulated genes) - (median value for proteins coded by down-regulated genes).
Circles and squares stand for hyperosmotic and hypoosmotic experiments, respectively.

```{r diffplot, fig.width=9, fig.height=8, fig.align = "center", echo = FALSE, pngquant = pngquant}
layout(matrix(c(0, 2, 1, 2, 1, 3, 0, 3), nrow = 2))
par(cex = 1, mar = c(4, 4, 1, 1), mgp = c(2.5, 1, 0))
pch <- ifelse(grepl("sorbitol", datasets), 1, 0)
diffplot(comptab1, labtext = NA, contour = FALSE, pch = pch)
diffplot(comptab2, c("pI", "GRAVY"), labtext = NA, contour = FALSE, pch = pch)
diffplot(comptab3, c("nAA", "MW"), labtext = NA, contour = FALSE, pch = pch)
```

In the table, values of &Delta;<i>Z</i><sub>C</sub>, &Delta;<i>n</i><sub>H<sub>2</sub>O</sub>, and &Delta;GRAVY are multiplied by 1000, values of &Delta;pI and &Delta;MW are multiplied by 100, and negative values are shown in bold.
Abbreviations:

  * <i>n</i><sub>down</sub> &ndash; number of down-regulated genes; <i>n</i><sub>up</sub> &ndash; number of up-regulated genes.

```{r xsummary, results="asis", echo = FALSE}
library(xtable)
out <- xsummary3(comptab1, comptab2, comptab3)
# round values and include dataset tags
tags <- sapply(sapply(strsplit(datasets, "="), "[", -1), paste, collapse = ";")
out <- cbind(out[, 1:2], tags = tags, out[, 3:19])
out[, 6:20] <- round(out[, 6:20], 4)
file <- paste0("yeast_stress_", basis, ".csv")
write.csv(out, file, row.names = FALSE, quote = 2)
```

## Data Sources

__a__. &ndash; __l__. [complete_dataset.txt](http://www-genome.stanford.edu/yeast_stress/data/rawdata/complete_dataset.txt) of @GSK+00, filtered to include genes with an expression ratio > log~2~(1.5) or < log~2~(1/1.5) at any time in the "1M sorbitol" or "Hypo-osmotic shock" experiments.

## References
