---
title: "Water in cancer (2020)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Water in cancer (2020)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

```{r options, include=FALSE}
options(width = 40)
```

```{r HTML, include=FALSE}
## some frequently used HTML expressions
ZC <- "<i>Z</i><sub>C</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
nAA <- "<i>n</i><sub>AA</sub>"
nO2 <- "<i>n</i><sub>O<sub>2</sub></sub>"
```

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
# read canprot's 'basis' option
basis <- getOption("basis")
```

This vignette contains plots from the paper by [<span title="Water as a reactant in the differential expression of proteins in cancer">Dick (2020)</span>](https://doi.org/10.1101/2020.04.09.035022).
This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version` and **[canprot](https://github.com/jedick/canprot)** `r packageDescription("canprot")$Version`.
The **`r basis`** basis species (`r basis.text(basis)`, H~2~O, O~2~) are used for the analysis of stoichiometric hydration state (`r nH2O`).

## Study overview (Figure 1)
```{r canH2O1, message = FALSE, fig.align='center', out.width='75%', fig.width = 7, fig.height = 7, pngquant = pngquant}
canH2O1()
```

## Median differences of protein length, `r nH2O` and `r ZC` for cell culture and cancer tissue (Figure 2)
```{r canH2O2, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant}
canH2O2()
```

## `r nH2O`-`r ZC` plots for TCGA and HPA datasets (Figure 3)
```{r canH2O3, message = FALSE, fig.align='center', out.width='100%', fig.width = 10, fig.height = 4, pngquant = pngquant}
canH2O3()
```

## Differentially expressed genes in aneuploid and osmotically shocked yeast cells (Figure 4)
```{r canH2O4, message = FALSE, results = "hide", fig.align='center', out.width='80%', fig.width = 5, fig.height = 4, pngquant = pngquant}
canH2O4()
```

## Mean differences and _p_-values for all datasets in each condition (Table 2)
Mean differences are followed by log~10~ of _p_-value in parentheses.
Mean differences are computed as (mean of median values for up-regulated proteins in each dataset) - (mean of median values for down-regulated proteins in each dataset).
The rows labelled "up" and "down" refer to (mean of median values for [up- or down-]regulated proteins in secreted proteins in hypoxia) - (mean of median values for [up-or down-]regulated proteins in whole-cell extracts in hypoxia).
```{r canH2OT2}
kable(canH2OT2(), align = "r")
```

# Supplementary Tables and Figures

## Stoichiometric matrix for amino acids with QEC basis species (Table S1)
```{r canH2OT1}
kable(canH2OT1())
```

## `r nO2`-`r ZC` and `r nH2O`-`r ZC` correlations using QEC basis species (Figure S1)
```{r canH2OS1, message = FALSE, fig.align='center', out.width='100%', fig.width = 7, fig.height = 3.5, pngquant = pngquant}
canH2OS1()
```

## HPA-TCGA scatterplots for `r ZC` and `r nH2O` (Figure S2)
```{r canH2OS2, message = FALSE, warning = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 4, pngquant = pngquant}
canH2OS2()
```
