---
title: "Water loss in evo-devo (2021)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Water loss in evo-devo (2021)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: JMDplots.bib
csl: elementa.csl
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

```{r options, include=FALSE}
options(width = 40)
```

```{r HTML, include=FALSE}
logfO2 <- "log<i>f</i><sub>O<sub>2</sub></sub>"
logaH2O <- "log<i>a</i><sub>H<sub>2</sub>O</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
```

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
# read canprot's 'basis' option
basis <- getOption("basis")
```

This vignette contains plots from the paper by [<span title="Redox potential linked to water loss from proteins in evolution and development">Dick (2021)</span>](evdevH2O.html).
This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`.

Phylostrata are from @TPPG17.
Consensus gene ages are from @LMM16.

## Compositional analysis of Trigos and Liebeskind datasets (Figure 1)
```{r evdevH2O1, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O1()
```

## Thermodynamic analysis of optimal `r logaH2O` and `r logfO2` for target proteins (Figure 2)
```{r evdevH2O2, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 3, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O2()
```

## Optimal `r logaH2O` and `r logfO2` and effective Eh for target proteins (Figure 3)
```{r evdevH2O3, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O3()
```

## Compositional and thermodynamic analysis of *B. subtilis* biofilm transcriptome and proteome (Figure 4)
Transcriptomic and proteomic data are from @FOK+20.
```{r evdevH2O4, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 4.5, pngquant = pngquant, dpi = 100}
evdevH2O4()
```

## Organismal water content and proteomic `r nH2O` for fruit fly development (Figure 5)
Whole-organism water content is from @CR66.
Stoichiometric hydration state of proteins is calculated in this study using proteomic data from @CBS+17.
```{r evdevH2O5, message = FALSE, results = "hide", out.width='100%', fig.width = 12, fig.height = 5, pngquant = pngquant, dpi = 100}
evdevH2O5()
```

## References
