---
title: "Thermodynamic model for water activity and redox potential in evolution and development (2022)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Thermodynamic model for water activity and redox potential in evolution and development (2022)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: JMDplots.bib
csl: elementa.csl
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 

## colorize messages 20171031
## adapted from https://gist.github.com/yihui/2629886#file-knitr-color-msg-rnw
color_block = function(color) {
  function(x, options) sprintf('<pre style="color:%s">%s</pre>', color, x)
}
knit_hooks$set(warning = color_block('magenta'), error = color_block('red'), message = color_block('blue'))
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
logfO2 <- "log<i>f</i><sub>O<sub>2</sub></sub>"
logaH2O <- "log<i>a</i><sub>H<sub>2</sub>O</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
ZC <- "<i>Z</i><sub>C</sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM. 2022. A thermodynamic model for water activity and redox potential in evolution and development. Submitted manuscript.
>
> Preprint available at:
>
> Dick JM. 2021. Redox potential linked to water loss from proteins in evolution and development. bioRxiv. doi: [10.1101/2021.01.29.428804](https://doi.org/10.1101/2021.01.29.428804)

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`, **[CHNOSZ](http://chnosz.net)** `r packageDescription("CHNOSZ")$Version`, and **[canprot](https://github.com/jedick/canprot)** `r packageDescription("canprot")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

To reduce running time, the plots in this vignette are made with 99 bootstrap replicates.
To reproduce the plots in the paper, the `boot.R` argument should be changed to 999.

## Protein length and chemical metrics for phylostratigraphic age groups (Figure 1)
```{r evdevH2O1, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O1(boot.R = 99)
```

**Data sources:** Phylostrata are from @TPPG17. Consensus gene ages are from @LMM16.

## MaximAct: Thermodynamic analysis of optimal `r logaH2O` and `r logfO2` for target proteins (Figure 2)
```{r evdevH2O2, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 3, out.extra='class="full-width"', pngquant = "--speed=1 --quality=25-50", dpi = 100}
evdevH2O2()
```

## Optimal `r logaH2O` and `r logfO2` and virtual Eh for target proteins (Figure 3)
```{r evdevH2O3, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O3()
```

**Data sources:** Plasma and subcellular redox potentials are from @VWR+13 and @JS15.

## Chemical metrics for and thermodynamic parameters with different background proteomes (Figure 4)
```{r evdevH2O4, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O4()
```

## Chemical and thermodynamic analysis of *B. subtilis* biofilm transcriptome and proteome (Figure 5)
```{r evdevH2O5, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 4.5, pngquant = pngquant, dpi = 100}
evdevH2O5(boot.R = 99)
```

**Data source:** Transcriptomic and proteomic data are from @FOK+21.

## Organismal water content, proteomic `r nH2O`, and optimal `r logaH2O` for fruit fly development (Figure 6)
```{r evdevH2O6, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 6, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O6(boot.R = 99)
```

**Data sources:**
**a** Whole-organism water content is from @CR66.
**b** Stoichiometric hydration state of proteins is calculated in this study using proteomic data from @CBS+17.
**d** `r ZC` and `r nH2O` of differentially expressed proteins in embryos and adult flies is calculated in this study using proteomic data from @FKL+19.


## Evolution of protein `r ZC` in eukaryotic lineages (Figure 7)
```{r evdevH2O7, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 6, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
evdevH2O7()
```

**Data sources:**
**a** Consensus gene ages are from @LMM16. Divergence times of human lineage are from @KSSH17.
**b** Amino acid compositions of homology groups for Pfam domains are from @JWN+20 and @JWN+21.

## Show total and unmapped number of genes in @TPPG17 phylostrata dataset
```{r TPPG17_stats, echo = FALSE}
file <- system.file("extdata/phylostrata/TPPG17.csv.xz", package = "canprot")
dat <- read.csv(file)
message(paste("Total number of genes:", nrow(dat)))
message("Total gene count in each phylostratum:")
table(dat$Phylostrata)
message(paste("Genes not mapped to UniProt:", sum(is.na(dat$Entry))))
message("Unmapped genes in each phylostratum:")
table(dat$Phylostrata[is.na(dat$Entry)])
```

## Example of protein chemical formula, formation reaction, equilibrium constant, and activity product
```{r LYSC_example}
LYSC_example()
```

## References
