---
title: "Evolutionary oxidation of proteins in Earth's history (2025)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Evolutionary oxidation of proteins in Earth's history (2025)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: JMDplots.bib
csl: elementa.csl
link-citations: true
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>
<script>
function ToggleDiv(ID) {
  var D = document.getElementById("D-" + ID);
  var B = document.getElementById("B-" + ID);
  if (D.style.display === "none") {
    // open the div and change button text
    D.style.display = "block";
    B.innerText = "Hide code";
  } else {
    // close the div and change button text
    D.style.display = "none";
    B.innerText = "Show code";
  }
}
</script>


```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 

## colorize messages 20171031
## adapted from https://gist.github.com/yihui/2629886#file-knitr-color-msg-rnw
color_block = function(color) {
  function(x, options) sprintf('<pre style="color:%s">%s</pre>', color, x)
}
knit_hooks$set(warning = color_block('magenta'), error = color_block('red'), message = color_block('blue'))
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
ZC <- "<i>Z</i><sub>C</sub>"
```
This vignette runs the functions to make the plots from the following manuscript:

> Dick JM et al. 2025. Evolutionary oxidation of proteins in Earth's history. Manuscript in preparation.

This vignette was compiled on `r Sys.Date()` with
**[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`,
**[canprot](https://github.com/jedick/canprot)** `r packageDescription("canprot")$Version`, and
**[CHNOSZ](http://chnosz.net)** `r packageDescription("CHNOSZ")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

## Genome-wide differences of oxidation state between two lineages of methanogens (Figure 1)
```{r genoGOE1, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant}
genoGOE_1()
```

**Data sources:** List of Class I and II methanogen species [@LL18], archaeal marker genes used in GTDB [@RCM+21], and protein abundance in *Methanococcus maripaludis* [@XHZ+06].

## Carbon oxidation state of proteins in eukaryotic gene age groups (Figure 2)
```{r genoGOE2, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 6, pngquant = pngquant}
genoGOE_2()
```

**Data sources:** Consensus gene ages [@LMM16], protein sequences [@Uni23], and divergence times [@KSC+22].

## Carbon oxidation state of ancestral reconstructed sequences and extant proteins (Figure 3)
```{r genoGOE3, message = FALSE, results = "hide", out.width='100%', fig.width = 9, fig.height = 6, pngquant = pngquant}
genoGOE_3()
```

**Data sources:** Thioredoxin [@PIZ+11], IPMDH [@CDY+25], Rubisco [@KHAB17], nitrogenase [@GMKK20].

## From carbon oxidation state to relative stability diagrams (Figure 4)
```{r genoGOE4, message = FALSE, results = "hide", out.width='100%', fig.width = 12, fig.height = 8, pngquant = pngquant}
genoGOE_4()
```

**Data sources:** Ancestral sequences of Rubisco [@KHAB17]; genome IDs for methanogens [@LL18], *Nitrososphaeria* [@RFH+19], and S-cycling organisms [@MCK+23].

## Composition reaction for ancestral Rubisco I/II/III (Reaction 1)

<button id="B-composition_calculation" onclick="ToggleDiv('composition_calculation')">Show code</button>
<div id="D-composition_calculation" style="display: none">
```{r composition_calculation, message = FALSE, results = "hide"}
# Setup basis species and swap in e- for O2
# NOTE: This sets pH to 7; changing it will change the charge (Z) and affinity below
basis("QEC+")
swap.basis("O2", "e-")

# Calculate Eh from pe
pe <- -basis()$logact[rownames(basis()) == "e-"]
Eh <- round(convert(pe, "Eh"), 3)

# Read amino acid compositions and add the first protein to CHNOSZ
fasta_file <- system.file("extdata/fasta/KHAB17.fasta", package = "canprot")
aa <- read_fasta(fasta_file)[1, ]
ip <- add.protein(aa)

# Run affinity() to calculate the affinity of the composition reaction for the first protein
affinity_out <- affinity(iprotein = ip)
affinity_protein <- affinity_out$values[[1]]

# Run protein.basis() to calculate the stoichiometry of the composition reaction
reaction_protein <- protein.basis(ip)

# NOTE: Affinity and reaction stoichiometry are already corrected for pH
# Next we construct the elemental formula of the uncharged protein ...
elements_protein <- protein.formula(ip)
# ... and make the pH correction
Z_protein <- ionize.aa(aa)[[1]]
elements_protein <- cbind(elements_protein, Z = 2)
elements_protein[, "H"] <- elements_protein[, "H"] + Z_protein

# Divide values by protein length and do some rounding
plength <- protein.length(ip)
affinity_residue <- round(affinity_protein / plength, 3)
reaction_residue <- round(reaction_protein / plength, 3)
elements_residue <- round(elements_protein / plength, 3)
# Format the elemental formula
formula_residue <- as.chemical.formula(round(elements_residue, 3))

# Print the results
print(paste("Eh:", Eh, "volt"))
print(paste("Elemental formula of protein per residue:", formula_residue))
print("Stoichiometry of composition reaction:")
print(reaction_residue)
print(paste("Affinity/2.303RT of composition reaction:", affinity_residue))
```
</div>

```{r composition_resuls, echo = FALSE}
print(paste("Eh:", Eh, "volt"))
print(paste("Elemental formula of protein per residue:", formula_residue))
print("Stoichiometry of composition reaction:")
print(reaction_residue)
print(paste("Affinity/2.303RT of composition reaction:", affinity_residue))
```

## References
