---
title: "Chemical metrics of estimated microbial community proteomes from 16S rRNA and RefSeq data (2021)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Chemical metrics of estimated microbial community proteomes from 16S rRNA and RefSeq data (2021)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: JMDplots.bib
csl: elementa.csl
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 

## colorize messages 20171031
## adapted from https://gist.github.com/yihui/2629886#file-knitr-color-msg-rnw
color_block = function(color) {
  function(x, options) sprintf('<pre style="color:%s">%s</pre>', color, x)
}
knit_hooks$set(warning = color_block('magenta'), error = color_block('red'), message = color_block('blue'))
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
logfO2 <- "log<i>f</i><sub>O<sub>2</sub></sub>"
logaH2O <- "log<i>a</i><sub>H<sub>2</sub>O</sub>"
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
ZC <- "<i>Z</i><sub>C</sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM, Tan J. 2021. Evidence for chemical links between redox conditions and microbial community composition from 16S rRNA and reference protein sequences. Submitted manuscript.
>
> Preprint available at:
>
> Dick JM, Tan J. 2021. Genomic evidence for a chemical link between redox conditions and microbial community composition. bioRxiv. doi: [10.1101/2021.05.31.446500](https://doi.org/10.1101/2021.05.31.446500)

<!--
The plots in this vignette are modified from the preprint in the following ways:

* All taxids are weighted equally (instead of by number of reference sequences) to calculate the mean amino acid composition for taxonomic groups.
* For subsets of samples, RDP counts are aggregated to calculate `r ZC` and `r nH2O` (instead of taking the mean of these values among samples).
* Removed dataset for NE Pennsylvania stream sediment.
* Added Fig. 6 and Fig. S1.
-->

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
# Tell the functions where the RDP Classifer output is stored (this is the default)
geo16Sdir = system.file("extdata/geo16S", package = "JMDplots")
options(chem16Sdir = geo16Sdir)
```

## Distinct chemical parameters of reference proteomes for major taxonomic groups (Figure 1)
```{r geo16S1, message = FALSE, warning = FALSE, results = "hide", out.width='100%', fig.width = 11, fig.height = 5, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
dat1 <- geo16S1()
```

**Data source:** NCBI Reference Sequence (RefSeq) database [@OWB+16]. **Numbered symbols**: (1) Methanococci, (2) Archaeoglobi, (3) Thermococci, (4) Halobacteria, (5) Nanohaloarchaea, (6) Clostridia.

Get the chemical metrics including `r ZC` for particular genera that are abundant in produced fluids of shale gas wells:

```{r taxon_metrics}
datadir <- system.file("extdata/chem16S", package = "JMDplots")
taxon_metrics <- read.csv(file.path(datadir, "taxon_metrics.csv"), as.is = TRUE)
subset(taxon_metrics, group %in% c("Halanaerobium", "Thermoanaerobacter"))
```

## Estimated community proteomes from different environments have distinct chemical signatures (Figure 2)
```{r geo16S2, message = FALSE, results = "hide", out.width='100%', fig.width = 9, fig.height = 7, out.extra='class="full-width"', pngquant = pngquant, dpi = 100}
dat2 <- geo16S2()
```

**Data sources:** Guerrero Negro mat [@HCW+13], Yellowstone hot springs [@BGPF13], Baltic Sea water [@HLA+16], Lake Fryxell mat [@JHM+16], Tibetan Plateau lakes [@ZLM+16], Manus Basin vents [@MPB+17], Qarhan Salt Lake soils [@XDZ+17], Black Sea water [@SVH+19].

## Lower carbon oxidation state is tied to oxygen depletion in water columns (Figure 3)
```{r geo16S3, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 9, pngquant = pngquant}
dat3 <- geo16S3()
```

**Data sources:** Black Sea [@SVH+19], Swiss lakes (Lake Zug and Lake Lugano) [@MZG+20], Eastern Tropical North Pacific (ETNP) [@GBL+15], Sansha Yongle Blue Hole [@HXZ+20], Ursu Lake [@BCA+21].

## Decreased carbon oxidation state of estimated proteomes for communities affected by shale gas extraction (Figure 4)
```{r geo16S4, message = FALSE, results = "hide", out.width='100%', fig.width = 9, fig.height = 6, pngquant = pngquant}
dat4 <- geo16S4()
```

**Data sources:** Northwestern Pennsylvania stream water and sediment [@UKD+18], Pennsylvania State Forests stream water in spring and fall [@MMA+20], Marcellus Shale [@CHM+14], Denver--Julesburg Basin [@HRR+18], Duvernay Formation [@ZLF+19].

## Contributions of classes to overall `r ZC` difference between oxidizing and reducing conditions (Figure 5)
```{r geo16S5, message = FALSE, results = "hide", out.width='100%', fig.width = 14, fig.height = 8, out.extra = 'class="full-width"', pngquant = pngquant}
geo16S5()
```

## Comparison of protein `r ZC` from metagenomic or metatranscriptomic data with estimates from 16S and reference sequences (Figure 6)
```{r geo16S6, message = FALSE, results = "hide", out.width='100%', fig.width = 7, fig.height = 5, pngquant = pngquant}
geo16S6()
```

**Data sources:** Guerrero Negro mat metagenome [@KRH+08], 16S [@HCW+13]; Bison Pool metagenome [@HRM+11], 16S [@SMS+12]; Eastern Tropical North Pacific metagenome [@GKG+15], 16S [@GBL+15]; Mono Lake metatranscriptome [@EH17], 16S [@EH18].

## RefSeq and 16S rRNA data processing outline (Figure S1)
```{r geo16S_S1, message = FALSE, results = "hide", out.width='100%', fig.width = 16, fig.height = 5, out.extra = 'class="full-width"', pngquant = pngquant}
geo16S_S1()
```

## Scatterplots of `r ZC` and `r nH2O` for 3443 bacterial and archaeal genera vs higher taxonomic levels (Figure S2)
```{r geo16S_S2, message = FALSE, results = "hide", out.width='100%', fig.width = 12, fig.height = 6, out.extra = 'class="full-width"', pngquant = pngquant}
geo16S_S2()
```

## Changes of abundance and chemical metrics at different taxonomic levels (Figure S3)
```{r geo16S_S3, message = FALSE, results = "hide", out.width='100%', fig.width = 10.5, fig.height = 7, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
geo16S_S3()
```

**Data sources:** Manus Basin vents [@MPB+17], Baltic Sea water [@HLA+16].


## Abundance and `r ZC` of classes in oxidizing and reducing conditions (Figure S4)
```{r geo16S_S4, message = FALSE, results = "hide", out.width='100%', fig.width = 9, fig.height = 12, pngquant = pngquant}
geo16S_S4()
```

## Source Data files
This code shows how the Source Data files for each of the figures is saved.
The `dat*` objects are created by running the code blocks above, but the following code block is not run in this vignette in order to avoid cluttering the working directory.
```{r SourceData, eval = FALSE}
write.csv(dat1, "Dataset_S1.csv", row.names = FALSE, quote = FALSE)
write.csv(dat2, "Dataset_S2.csv", row.names = FALSE, quote = FALSE)
write.csv(dat3, "Dataset_S3.csv", row.names = FALSE, quote = FALSE)
write.csv(dat4, "Dataset_S4.csv", row.names = FALSE, quote = FALSE)
```

## References
