---
title: "Carbon oxidation state from metagenomic data for redox gradients (2019)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Carbon oxidation state from metagenomic data for redox gradients (2019)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

```{r HTML, include=FALSE}
## some frequently used HTML expressions
ZC <- "<i>Z</i><sub>C</sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM, Yu M, Tan J, Lu A. 2019. Changes in carbon oxidation state of metagenomes along geochemical redox gradients. *Frontiers in Microbiology* **10**: 120. doi: [10.3389/fmicb.2019.00120](https://doi.org/10.3389/fmicb.2019.00120)

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version` and **[CHNOSZ](http://chnosz.net)** `r packageDescription("CHNOSZ")$Version`.

```{r JMDplots, message = FALSE, results = "hide"}
library(JMDplots)
```

## General characteristics of `r ZC` of DNA, RNA, and proteins (Figure 1)
```{r gradox1, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant}
gradox1()
```

## Selected plots of metagenomic DNA, RNA, and protein `r ZC` (Figure 2)
```{r gradox2, message = FALSE, fig.align='center', out.width='100%', fig.width = 9, fig.height = 9, pngquant = pngquant}
gradox2()
```

## All plots of DNA and RNA `r ZC` (Figure S1)
```{r gradoxS1, message = FALSE, fig.align='center', out.width='100%', fig.width = 6.5, fig.height = 10, pngquant = pngquant}
mout <- gradoxS1()
```

## All plots of protein `r ZC` (Figure S2)
```{r gradoxS2, message = FALSE, fig.align='center', out.width='100%', fig.width = 6.5, fig.height = 10, pngquant = pngquant}
pout <- gradoxS2()
```

## `r ZC` of proteins vs DNA (metagenomes and metatranscriptomes) (Figure 3)
```{r gradox3, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 5, pngquant = pngquant}
gradox3(mout, pout)
```

## Thermodynamic calculations of relative stabilities along redox gradients (Figure 4)
```{r gradox4, message = FALSE, fig.align='center', out.width='100%', fig.width = 7.5, fig.height = 6.5, pngquant = pngquant}
gradox4(mout, pout)
```

## `r ZC` of reads classified to selected abundant species (Figure 5)
```{r gradox5, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 5, pngquant = pngquant}
gradox5()
```

## `r ZC` of reads classified to selected abundant species (including deeper samples) (Figure S3)
```{r gradoxS3, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 5, pngquant = pngquant}
gradox5(maxdepth = NULL)
```

