---
title: "Signatures of salinity gradients in the chemical composition of proteins (2020)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Signatures of salinity gradients in the chemical composition of proteins (2020)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: JMDplots.bib
csl: elementa.csl
link-citations: true
---

```{r setup, include=FALSE}
library(knitr)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

```{r options, include=FALSE}
options(width = 40)
```

```{r HTML, include=FALSE}
## some frequently used HTML expressions
ZC <- "<i>Z</i><sub>C</sub>"
nH2O <- "<i>n</i>H<sub>2</sub>O"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM, Yu M, Tan J. 2020. Uncovering chemical signatures of salinity gradients through compositional analysis of protein sequences. *Biogeosciences* **17**(23): 6145--6162. doi: [10.5194/bg-17-6145-2020](https://doi.org/10.5194/bg-17-6145-2020)

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version` and **[canprot](https://github.com/jedick/canprot)** `r packageDescription("canprot")$Version`.

```{r refseq, echo = FALSE}
# Get RefSeq version 20210527
file <- system.file("extdata/RefSeq/README.txt", package = "chem16S")
release <- substr(strsplit(readLines(file, n=2)[2], "release ")[[1]][2], 1, 3)
```

For Figure 3, the RefSeq database release `r release` is used for the proteomes of *nif*-bearing organisms (release 201 was used in the paper).

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

The QEC basis species (glutamine, glutamic acid, cysteine, H~2~O, O~2~) are used for the analysis of stoichiometric hydration state (`r nH2O`).

## Number of reactions for each amino acid in _E. coli_ metabolic reconstruction
Reaction equations used for this calculation are from <a href="https://doi.org/10.1038/msb4100155" title="A genome-scale metabolic reconstruction for Escherichia coli K-12 MG1655 that accounts for 1260 ORFs and thermodynamic information">Feist et al. (2007)</a>.
```{r gradH2O0, message = FALSE, echo = 1}
gradH2O0()
```

## Comparison of different sets of basis species (Figure 1)
```{r gradH2O1, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 4.3, pngquant = pngquant}
gradH2O1()
```

## Schematic of `r nH2O` and `r ZC` calculations (Figure 2)
```{r gradH2O2, message = FALSE, fig.align='center', out.width='75%', fig.width = 5, fig.height = 2.5, pngquant = pngquant}
gradH2O2()
```

## `r nH2O`-`r ZC` scatterplots for redox gradients and the Baltic Sea (Figure 3)
```{r gradH2O3, message = FALSE, fig.align='center', out.width='100%', fig.width = 12, fig.height = 5.6, pngquant = pngquant}
gradH2O3()
```

**Data sources:** Bison Pool [BP; @HRM+11; @SMS+12]; diffuse submarine vents [DV; @RRM+16; @FLBH18]; Guerrero Negro microbial mat [GN; @KRH+08]; nitrogenase-bearing organisms [NF; @PCF+18]; Baltic Sea [@DLY+14].

## `r nH2O` for Baltic Sea metagenome and metatranscriptome in different size fractions (Figure 4)
```{r gradH2O4, message = FALSE, fig.align='center', out.width='100%', fig.width = 6, fig.height = 3.5, pngquant = pngquant}
gradH2O4()
```

**Data sources:** metagenome [@DLY+14] and metatranscriptome [@ASD+16].

## `r nH2O`-`r ZC` for freshwater, marine, and hypersaline environments (Figure 5)
```{r gradH2O5, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 5, pngquant = pngquant}
gradH2O5()
```

**Data sources:** Amazon river [@SFD+15] and plume [@SZD+14]; freshwater and marine metagenomes [@EZM+14]; hypersaline datasets Kulunda Steppe [KS; @VGR+16], Santa Pola [SP; @GPF+11; @FGM+13], San Francisco South Bay [SB; @KBW+18].

## `r nH2O`-`r ZC` and GRAVY-pI plots for Baltic Sea and Rodriguez-Brito et al. data (Figure 6)
```{r gradH2O6, message = FALSE, fig.align='center', out.width='75%', fig.width = 8, fig.height = 8, pngquant = pngquant}
gradH2O6()
```

**Data sources:** @DLY+14 and @RLW+10.

## Differential gene and protein expression; time-course experiments and NaCl or organic solutes (Figure 7)
```{r gradH2O7, message = FALSE, fig.align='center', out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant}
gradH2O7()
```

**Data sources:** *Panel (a):* a--c: @KKG+14; d--f: @SLM+14; g--i: @FRH+15; j--k: @HLL17; l--n: @QHT+13.
*Panel (b):* A--B: @KSA+02; C--D: @HZP+05; E--F: @KLB+15; G--J: @FRH+15; K--L: @SBB+09; M--N: @WGB+13.
*Panels (c) to (f):* See the vignettes `JMDplots::mkvig("osmotic_bact")` and `JMDplots::mkvig("osmotic_halo")`. Links to static vignettes: [osmotic_bact](../diffexpr/osmotic_bact.html) and [osmotic_halo](../diffexpr/osmotic_halo.html).

## References
