---
title: "Water and oxygen in the genomic adaptation of human microbiota (2023)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Water and oxygen in the genomic adaptation of human microbiota (2023)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: [orp16S.bib, JMDplots.bib]
csl: elementa.csl
link-citations: true
---

```{r setup, include=FALSE}
library(knitr)
## Use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# In case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
nH2O <- "<i>n</i><sub>H<sub>2</sub>O</sub>"
nO2 <- "<i>n</i><sub>O<sub>2</sub></sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM. 2023. Water and oxygen in the genomic adaptation of human microbiota. Submitted manuscript.
> 
> A preprint entitled "Two dimensions of chemical variation of the human microbiome across body sites and in COVID-19 patients" is available at bioRxiv. doi: [10.1101/2023.02.12.528246](https://doi.org/10.1101/2023.02.12.528246)

This vignette was compiled on `r Sys.Date()` with
**[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`,
**[chem16S](https://github.com/jedick/chem16S)** `r packageDescription("chem16S")$Version`, and
**[canprot](https://github.com/jedick/canprot)** `r packageDescription("canprot")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

## Compatibility of inferences from shotgun metagenomes and community reference proteomes (Figure 1)
```{r microhum_1, message = FALSE, out.width='100%', fig.width = 10, fig.height = 6, pngquant = pngquant, dpi = 100}
microhum_1()
```

**Data source:** @HMP12.

## Chemical metrics are broadly different among genera and are similar between GTDB and low-contamination genomes from MGnify (Figure 2)
```{r microhum_2, message = FALSE, out.width='100%', fig.width = 12, fig.height = 7, pngquant = pngquant, dpi = 100}
microhum_2()
```

## Chemical variation of microbial proteins across body sites (multi-omics comparison) (Figure 3)
```{r microhum_3, message = FALSE, out.width='100%', fig.width = 7, fig.height = 6, pngquant = pngquant, dpi = 100}
microhum_3()
```

**Data sources:** **(A)** @BPB+21; **(B)** See Table 1 in manuscript; **(C)** @LLZ+21, @CZH+22, @ZZL+20; **(D)** @TWC+22, @MLL+17, @JZW+22, @GNT+21.

## Differences of chemical metrics between controls and COVID-19/IBD patients (Figure 4)
```{r microhum_4, message = FALSE, out.width='100%', fig.width = 13, fig.height = 12, pngquant = pngquant, dpi = 100}
microhum_4()
```

**Data sources:** **(A)** See Table 1 in manuscript; **(B)** MAGs [@KWL22], based on data from @YZL+21; **(C)** @HZX+21; **(D)** @GPM+22; **(E)** See Table 1 in manuscript; **(F)** @LAA+19.

## Differences of relative abundances of genera between controls and patients (Figure 5)
```{r microhum_5, message = FALSE, out.width='100%', fig.width = 10, fig.height = 8.5, pngquant = pngquant, dpi = 100}
microhum_5()
```

## Oxygen tolerance of genera in body sites, COVID-19, and IBD (Figure 6)
```{r microhum_6, message = FALSE, out.width='100%', fig.width = 12, fig.height = 9, pngquant = pngquant, dpi = 100}
microhum_6()
```

**Data sources:** **(A)--(B)** *List of Prokaryotes according to their Aerotolerant or Obligate Anaerobic Metabolism* [@MR18] with modifications made in this study; **(C)** See Table 1 in manuscript.

## Amount of putative human DNA removed from HMP metagenomes in screening step (Figure S1)
```{r microhum_S1, message = FALSE, out.width='100%', fig.width = 10, fig.height = 8, pngquant = pngquant, dpi = 100}
microhum_S1()
```

## Changes of chemical metrics for microbiomes associated with viral inactivation (Figure S2)
```{r microhum_S2, message = FALSE, out.width='100%', fig.width = 6, fig.height = 4, pngquant = pngquant, dpi = 100}
microhum_S2()
```

**Data source:** @BPB+21.

## Differences of `r nO2` and `r nH2O` between obligate anaerobic and aerotolerant genera (Figure S3)
```{r microhum_S3, message = FALSE, out.width='75%', fig.align='center', fig.width = 7, fig.height = 4, pngquant = pngquant, dpi = 100}
microhum_S3()
```

**Data source:** *List of Prokaryotes according to their Aerotolerant or Obligate Anaerobic Metabolism* [@MR18] with modifications made in this study.

## References
