---
title: "Chemical adaptation of bacterial communities to redox potential (2022)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Chemical adaptation of bacterial communities to redox potential (2022)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: [orp16S.bib, JMDplots.bib]
csl: elementa.csl
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 

## colorize messages 20171031
## adapted from https://gist.github.com/yihui/2629886#file-knitr-color-msg-rnw
color_block = function(color) {
  function(x, options) sprintf('<pre style="color:%s">%s</pre>', color, x)
}
knit_hooks$set(warning = color_block('magenta'), error = color_block('red'), message = color_block('blue'))
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
ZC <- "<i>Z</i><sub>C</sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM, Meng D. 2022. Estimated proteomes of bacterial communities are chemically adapted to redox conditions at a global scale. Submitted manuscript.
>
> Preprint available at bioRxiv. doi: [10.1101/2021.10.12.464155](https://doi.org/10.1101/2021.10.12.464155)

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

## Outline of methods (Figure 1)
```{r orp16S1, message = FALSE, results = "hide", out.width='75%', fig.align='center', fig.width = 6, fig.height = 4, pngquant = pngquant}
orp16S1()
```

## Chemical depth profiles in Winogradsky columns (Figure 2)
```{r orp16S2, message = FALSE, results = "hide", out.width='80%', fig.align='center', fig.width = 7, fig.height = 5, pngquant = pngquant}
orp16S2()
```

**Data sources:** ORP and pH for Eh7 calculation [@DFYS19]; 16S rRNA gene sequences for `r ZC` calculation [@RBW+14].

## Sample locations on world map (Figure 3)
```{r orp16S3, message = FALSE, results = "hide", out.width='100%', fig.width = 26, fig.height = 15, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S3()
```

**Data sources:** coastlineWorld dataset in R package oce [@KR21]; shapefiles for North American Great Lakes [@USGS10].

## Linear regressions between Eh7 and `r ZC` at local scales (Figure 4)
```{r orp16S4, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant, dpi = 100}
orp16S4()
```

**Data sources (a):** Daya Bay [@WHLH21], Bay of Biscay [@LMBA21], Hunan Province [@MLL+19].

**Data sources (c):** **1** acidic and **2** circumneutral to alkaline New Zealand hot springs [@PCL+18], **3** Dallol Geothermal Area [@BMJ+19], **4** Eastern Tibetan Plateau [@GWS+20], **5** Uzon Caldera [@PBU+20], **6** Southern Tibetan Plateau [@MWY+21].

## Summary of regression slopes for all datasets (Table 2)
```{r orp16S_T2}
colnames <- c("Number", "% Positive", "% Negative")
kable(orp16S_T2(), col.names = rep(colnames, 2)) %>%
  add_header_above(c(" " = 1, "Bacteria" = 3, "Archaea" = 3))
```

**Note:** Positive and negative refer to slopes of linear regressions that are > 0.01 V^-1^ and < -0.01 V^-1^, respectively.
Datasets for which the regression slopes have a smaller magnitude are not counted in either column.

## Linear regressions between Eh7 and `r ZC` at a global scale (Figure 5)
```{r orp16S5, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 7, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
global.slopes <- orp16S5()
```

## Eh-pH diagram (Figure S1)
```{r orp16S_S1, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant}
orp16S_S1()
```

**Data source:** Outline is from @BKM60.

## Regression slopes from local to global scale (Figure S3)
```{r orp16S_S3, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 3, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S_S3(global.slopes)
```

## Comparison of Eh7, Eh, and O~2~ as predictors of carbon oxidation state (Figure S4)
```{r orp16S_S4, message = FALSE, results = "hide", out.width='100%', fig.width = 6, fig.height = 8, pngquant = pngquant, dpi = 100}
orp16S_S4()
```

## References
