---
title: "Chemical adaptation of reference proteomes for bacterial communities (2022)"
output: 
  rmarkdown::html_vignette:
    mathjax: null
vignette: >
  %\VignetteIndexEntry{Chemical adaptation of reference proteomes for bacterial communities (2022)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: [orp16S.bib, JMDplots.bib]
csl: elementa.csl
---

<style>
/* https://gomakethings.com/how-to-break-an-image-out-of-its-parent-container-with-css/ */
@media (min-width: 700px) {
  .full-width {
    left: 50%;
    margin-left: -50vw;
    margin-right: -50vw;
    max-width: 100vw;
    position: relative;
    right: 50%;
    width: 100vw;
  }
}
@media (min-width: 900px) {
  .full-width {
    left: 50vw; /* fallback if needed */
    left: calc(50vw - 100px);
    width: 900px;
    position: relative;
  }
}
</style>

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
## use pngquant to reduce size of PNG images
knit_hooks$set(pngquant = hook_pngquant)
pngquant <- "--speed=1 --quality=0-25"
# in case pngquant isn't available
if (!nzchar(Sys.which("pngquant"))) pngquant <- NULL 

## colorize messages 20171031
## adapted from https://gist.github.com/yihui/2629886#file-knitr-color-msg-rnw
color_block = function(color) {
  function(x, options) sprintf('<pre style="color:%s">%s</pre>', color, x)
}
knit_hooks$set(warning = color_block('magenta'), error = color_block('red'), message = color_block('blue'))
```

```{r options, include=FALSE}
options(width = 80)
```

```{r HTML, include=FALSE}
ZC <- "<i>Z</i><sub>C</sub>"
```

This vignette runs the code to make the plots from the following paper:

> Dick JM, Meng D. 2022. Reference proteomes for bacterial communities reveal chemical adaptation to redox conditions at a global scale. Submitted manuscript.
>
> Preprint available at bioRxiv. doi: [10.1101/2021.10.12.464155](https://doi.org/10.1101/2021.10.12.464155)

This vignette was compiled on `r Sys.Date()` with **[JMDplots](https://github.com/jedick/JMDplots)** `r packageDescription("JMDplots")$Version` and **[chem16S](https://github.com/jedick/chem16S)** `r packageDescription("chem16S")$Version`.

```{r library, message = FALSE, results = "hide"}
library(JMDplots)
```

## Methods overview and chemical depth profiles in Winogradsky columns (Figure 1)
```{r orp16S_1, message = FALSE, results = "hide", out.width='100%', fig.width = 11.25, fig.height = 4, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S_1()
```

**Data sources:** **(b)** ORP and pH [@DFYS19] to calculate Eh7. **(c)** 16S rRNA gene sequences [@RBW+14] to calculate `r ZC`.

## Sample locations on world map and Eh-pH diagram (Figure 2)
```{r orp16S_2, message = FALSE, results = "hide", out.width='100%', fig.width = 26, fig.height = 23, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S_2()
```

**Data sources:** **(a)** coastlineWorld dataset in R package oce [@KR21]; shapefiles for North American Great Lakes [@USGS10]. **(b)** Outline is from @BKM60.

## Associations between Eh7 and `r ZC` at local scales (Figure 3)
```{r orp16S_3, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 6, pngquant = pngquant, dpi = 100}
orp16S_3()
```

**Data sources:** **(a)** Daya Bay [@WHLH21a], Bay of Biscay [@LMBA21], Hunan Province [@MLL+19]. **(c)** **1** acidic and **2** circumneutral to alkaline New Zealand hot springs [@PCL+18], **3** Eastern Tibetan Plateau [@GWS+20], **4** Uzon Caldera [@PBU+20], **5** Southern Tibetan Plateau [@MWY+21].

## Tally of regression slopes and binomial *p*-values (Table 2)
```{r orp16S_T2}
colnames <- c("Ntot", "Npos", "Nneg", "p")
kable(orp16S_T2(), col.names = rep(colnames, 2)) %>%
  add_header_above(c(" " = 1, "Bacteria" = 4, "Archaea" = 4))
```

## Tally of regression slopes with same sign of minimum and maximum values in 95% confidence interval (Table S3)
```{r orp16S_T2_S3}
colnames <- c("Ntot", "Npos", "Nneg", "p")
kable(orp16S_T2(samesign = TRUE), col.names = rep(colnames, 2)) %>%
  add_header_above(c(" " = 1, "Bacteria" = 4, "Archaea" = 4))
```

## Associations between Eh7 and `r ZC` at a global scale (Figure 4)
```{r orp16S_4, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 7, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
global.slopes <- orp16S_4()
```

## Comparison of 16S-based community reference proteomes with metaproteomes (Figure 5)
```{r orp16S_5, message = FALSE, results = "hide", out.width='100%', fig.align='center', fig.width = 8, fig.height = 7, pngquant = pngquant, dpi = 100}
orp16S_5()
```

**Data sources:** 16S rRNA gene sequences [@WHLH21], metaproteomes [@HWLH22].

## Most abundant genera at low and high Eh7 in geothermal and hyperalkaline areas (Figure 6)
```{r orp16S_6, message = FALSE, results = "hide", out.width='100%', fig.align='center', fig.width = 8.5, fig.height = 5, pngquant = pngquant, dpi = 100}
orp16S_6()
```

Bold and normal text are used for archaea and bacteria, respectively.

## Regression slopes from local to global scale (Figure S4)
```{r orp16S_S4, message = FALSE, results = "hide", out.width='100%', fig.width = 8, fig.height = 3, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S_S4(global.slopes)
```

## Comparison of Eh7, Eh, and O~2~ as predictors of carbon oxidation state (Figure S5)
```{r orp16S_S5, message = FALSE, results = "hide", out.width='100%', fig.width = 6, fig.height = 8, pngquant = pngquant, dpi = 100}
orp16S_S5()
```

## Global analysis including only datasets using 515F/806R primers from the Earth Microbiome Project (EMP) (Figure S6)
```{r orp16S_S6, message = FALSE, results = "hide", out.width='100%', fig.width = 10, fig.height = 7, out.extra = 'class="full-width"', pngquant = pngquant, dpi = 100}
orp16S_4(EMP_primers = TRUE)
```

## Genus-level RDP classifications mapped to NCBI taxonomy (in-text calculation)
```{r mapperc}
# Read Dataset S3 (created by orp16S_D3())
dat <- read.csv(system.file("extdata/orp16S/Dataset_S3.csv", package = "JMDplots"))
perc_bac <- round(mean(subset(dat, lineage == "Bacteria")$mapperc))
perc_arc <- round(mean(subset(dat, lineage == "Archaea")$mapperc))
message(paste0(perc_bac, "% for bacteria, ", perc_arc, "% for archaea"))
```

## References
